libraries/abstractions/common_io/include/iot_sdio.h

```diff
@@ -0,0 +1,709 @@
+ *          Here we define SDIO driver as a protocol layer driver which builds on top of SDMMC
+ *          host driver. SDMMC host driver operates at physical layer and provide APIs to send
+ *          commands to the slave device and receive command responses, send and receive data,
+ *          and handle error conditions on the bus.
+ *
+ *          SD/SDIO/MMC driver can be built on top of SDMMC host. This interface only define
+ *          the APIs for SDIO applications, the interface can be expanded to support SD/MMC either
+ *          by expanding the APIs in this file or adding separate files for SD/MMC
+ *
+ * @design consideration
+ *          SDIO HAL APIs performs protocol layer tasks between sdio host and sdio card.
+ *          Some APIs perform a single task and some APIs perform a sequence of tasks.
+ *          For examples, iot_sdio_io_read_direct() API performs a single task of reading
+ *          one byte from card register, and iot_sdio_card_connect() API performs all tasks
+ *          required by card initialization as specified in section 3.1.2 in "SD Specification
+ *          Part E1 SDIO".
+ *
+ *          Upon calling iot_sdio_open() API, the user is a given a pxSdioHandle, an abstract
+ *          reference to the SDIO card slot that the user is interested in. Thereafter, user
+ *          uses this handle to call other APIs to perform tasks associated with the target
+ *          card slot. For example, iot_sdio_check_card_presence_status(pxSdioHandle, true)
+ *          checks sdio card presence, and iot_sdio_io_write_extended(pxSdioHandle)
+ *          writes data to sdio card in the slot, etc.
+ *
+ * @limitation
+ *          SDIO HAL APIs are intended to be used in embedded systems where each sdio host
+ *          controller is connected to either a single hot swappable sido card slot or multiple
+ *          eSDIO devices in a Shared Bus Configuration (Refer to Section 7.3 of "SD Specification
+ *          Part E1 SDIO").
+ *
+ *          These APIs are not suitable for systems where a single sdio host controller supports
+ *          multiple hot swappable slots. With multiple hot swappable slots available on a single
+ *          host controller, it is not possible pre-determine which card slot sdio device will be
+ *          inserted, and know which card slot handle to request. Since this type of host controller
+ *          is very rare even in non-embedded world, it is not worth to add complexity to SDIO HAL
+ *          APIs to support it.
+ */
+
+#ifndef _IOT_SDIO_H_
+#define _IOT_SDIO_H_
+
+#include <stdint.h>
+#include <stdbool.h>
+
+
+/**
+ * @defgroup iot_sdio SDIO HAL APIs
+ * @{
+ */
+
+/**
+ * @brief The return codes for the functions in SDIO
+ */
+ /*!< SDIO operation completed successfully. */
+#define IOT_SDIO_SUCCESS                             ( 0 )
+/*!< SDIO operation failed. */
+#define IOT_SDIO_FAIL                                ( 1 )
```

Comment:
If this is specific to ioctl, can we rename it to IOT_SDIO_IOCTL_FAIL

Reply:
renamed.